File: il4il_sharp/src/Il4ilSharp/Interop/ErrorHandling.cs
namespace Il4ilSharp.Interop;

using System;
using System.Buffers;
using System.Text;
using Il4ilSharp.Interop.Native;

internal unsafe static class ErrorHandling {
    internal static void Throw(Error.Opaque* error) {
        if (error != null) {
            using ErrorHandle handle = new ErrorHandle(error);
            throw new ErrorHandlingException(handle.ToString());
        }
    }
}

File: il4il_sharp/test/Il4ilSharp.Properties/ErrorHandling.fs
module Il4ilSharp.Properties.ErrorHandling

open Il4ilSharp.Interop

open Expecto

[<Tests>]
let tests =
    testList "error handling" [
        testProperty "handle from string returns original string" <| fun (msg: string) ->
            let expected = if msg = null then System.String.Empty else msg
            use handle = new ErrorHandle(expected)
            Expect.equal (handle.ToString()) expected "error messages should be the same"
    ]
